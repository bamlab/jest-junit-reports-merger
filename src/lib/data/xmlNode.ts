export class XMLNode {
  constructor(public kind: string, public attributes: Record<string, string | number>) {}

  changeKind(kind: string) {
    this.kind = kind;
  }

  updateAttributes(attributes: Record<string, string>) {
    Object.entries(attributes).forEach(([value, key]) => {
      this.attributes[key] = value;
    });
  }
}

export class XMLComponent extends XMLNode {
  constructor(public kind: string, public attributes: Record<string, string | number>, public children: XMLNode[]) {
    super(kind, attributes);
  }
}

export class XMLElement extends XMLNode {
  constructor(public kind: string, public attributes: Record<string, string | number>, public content: string) {
    super(kind, attributes);
  }
}
