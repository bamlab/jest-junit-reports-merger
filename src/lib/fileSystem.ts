import glob from 'fast-glob';

export class FileSystem {
  constructor(private readonly root: string) {}

  listFilesPath(globPattern: string): string[] {
    return glob.sync(globPattern, { cwd: this.root, dot: true });
  }
}
