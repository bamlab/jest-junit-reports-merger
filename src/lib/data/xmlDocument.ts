import { XMLComponent, XMLNode } from './xmlNode';

export class XMLDocument {
  constructor(public nodes: XMLNode[]) {}

  findTopLevelComponent(kind: string): XMLComponent | undefined {
    return this.nodes.find((node) => node.kind === kind && node instanceof XMLComponent) as XMLComponent;
  }
}
