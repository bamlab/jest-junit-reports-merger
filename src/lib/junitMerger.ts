import { XMLDocument } from './data/xmlDocument';
import { XMLComponent, XMLElement } from './data/xmlNode';
import { XML } from './xml';

export class JUnitMerger {
  junitDocument: XMLDocument;

  testSuiteRecord: Record<string, XMLComponent> = {};

  constructor(private xmlModule: XML = new XML()) {
    const header = new XMLComponent('?xml', { version: '1.0', encoding: 'UTF-8' }, [new XMLElement('#text', {}, '')]);
    this.junitDocument = new XMLDocument([header]);
  }

  ingest(junitFilesContents: string[]) {
    for (const fileContent of junitFilesContents) {
      const newJunitDocument = this.xmlModule.parse(fileContent);
      const newTestsuites = newJunitDocument.findTopLevelComponent('testsuites');
      if (!newTestsuites) {
        continue;
      }
      this.appendTestsuites(newTestsuites);
    }
  }

  private appendTestsuites(newTestsuites: XMLComponent) {
    const name = newTestsuites.attributes.name;

    if (!name) {
      this.testSuiteRecord[Date.toString()] = newTestsuites;
    } else if (name in this.testSuiteRecord) {
      this.testSuiteRecord[name].children.push(...newTestsuites.children);
      const metrics = this.sumMetrics([this.testSuiteRecord[name].attributes, newTestsuites.attributes]);
      this.testSuiteRecord[name].attributes = metrics;
    } else {
      this.testSuiteRecord[name] = newTestsuites;
    }
  }

  private sumMetrics(metrics: Record<string, string | number>[]) {
    const result: Record<string, number> = {};

    for (const metric of metrics) {
      for (const [key, value] of Object.entries(metric)) {
        if (typeof value !== 'number') {
          continue;
        }
        if (key in result) {
          result[key] += value;
        } else {
          result[key] = value;
        }
      }
    }
    

    return result;
  }

  print() {
    let numberOfTestsuites = Object.keys(this.testSuiteRecord).length;

    let testsuites;
    switch (numberOfTestsuites) {
      case 0:
        break;
      case 1:
        testsuites = this.testSuiteRecord[Object.keys(this.testSuiteRecord)[0]];
        break;
      default:
        const attibutesList = Object.values(this.testSuiteRecord).map((testsuite) => testsuite.attributes);
        const resultingAttributes = {
          name: 'Jest Tests',
          ...this.sumMetrics(attibutesList),
        };
        Object.values(this.testSuiteRecord).forEach((testsuite) => testsuite.changeKind('testsuite'));
        testsuites = new XMLComponent('testsuites', resultingAttributes, Object.values(this.testSuiteRecord));
    }

    if (testsuites) {
      this.junitDocument.nodes.push(testsuites);
    }
    return this.xmlModule.print(this.junitDocument);
  }
}
