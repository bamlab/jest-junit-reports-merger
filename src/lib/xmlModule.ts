import { XMLParser, XMLBuilder, XmlBuilderOptions, X2jOptions } from 'fast-xml-parser';
import { XMLDocument } from './data/xmlDocument';
import { XMLComponent, XMLElement, XMLNode } from './data/xmlNode';

type ParsedXMLAttribute<AttributeName extends string> = `@_${AttributeName}`;

type ParsedXMLAttributes<AttributeName extends string> = Record<ParsedXMLAttribute<AttributeName>, string>;

type ParsedXMlElementPartName<NodeName extends string> = {
  [key in NodeName]: ParsedXMLNode<string>[];
};

type ParsedXMlComponentPartName<NodeName extends string> = {
  [key in NodeName]: string;
};

type ParsedXMlNodePartAttributes<AttributeName extends string> = {
  ':@': ParsedXMLAttributes<AttributeName>;
};

type ParsedXMLElement<NodeName extends string, AttributeName extends string = string> =
  | ParsedXMlElementPartName<NodeName> & Partial<ParsedXMlNodePartAttributes<AttributeName>>;

type ParsedXMLComponent<NodeName extends string, AttributeName extends string = string> =
  | ParsedXMlComponentPartName<NodeName> & Partial<ParsedXMlNodePartAttributes<AttributeName>>;

type ParsedXMLNode<NodeName extends string, AttributeName extends string = string> =
  | ParsedXMLElement<NodeName, AttributeName>
  | ParsedXMLComponent<NodeName, AttributeName>;

const PARSED_XML_ATTRIBUTES_KEY = ':@';

const XML_OPTIONS: Partial<XmlBuilderOptions> & Partial<X2jOptions> = {
  ignoreAttributes: false,
  attributeNamePrefix: '@_',
  format: true,
  preserveOrder: true,
};

export class XML {
  constructor(
    private xmlParser: XMLParser = new XMLParser(XML_OPTIONS),
    private xmlBuilder: XMLBuilder = new XMLBuilder(XML_OPTIONS)
  ) {}

  parse = (xml: string): XMLDocument => {
    const parsedXmlNodes = this.xmlParser.parse(xml) as ParsedXMLNode<string>[];
    const nodes = parsedXmlNodes.map(this.parseNode);
    return new XMLDocument(nodes);
  };

  private parseNode = (node: ParsedXMLNode<string>): XMLNode => {
    const nodeKind = Object.keys(node).filter((key) => key !== PARSED_XML_ATTRIBUTES_KEY)[0];

    const parsedAttributes = this.parseAttributes(node[PARSED_XML_ATTRIBUTES_KEY]);

    const nodeChildren = node[nodeKind];

    if (typeof nodeChildren === 'string') {
      return new XMLElement(nodeKind, parsedAttributes, nodeChildren);
    } else {
      const parsedChildren = nodeChildren.map(this.parseNode);
      return new XMLComponent(nodeKind, parsedAttributes, parsedChildren);
    }
  };

  private parseAttributes = (attributes: Record<string, string> | undefined): Record<string, string> => {
    if (!attributes) return {};
    return Object.entries(attributes).reduce((parsedAttributes, [key, value]) => {
      const attributeName = key.replace(/^@_/, '') as string;

      let parsedValue;
      if (['version'].includes(attributeName)) {
        parsedValue = value;
      } else if (value.match(/^[0-9]+$/)) {
        parsedValue = parseInt(value, 10);
      } else if (value.match(/^[0-9]+\.[0-9]+$/)) {
        parsedValue = parseFloat(value);
      } else {
        parsedValue = value;
      }

      return {
        ...parsedAttributes,
        [attributeName]: parsedValue,
      };
    }, {});
  };

  print = (xmlDocument: XMLDocument): string => {
    const xmlNodes = xmlDocument.nodes.map(this.printNode);
    return this.xmlBuilder.build(xmlNodes);
  };

  private printNode = (xmlNode: XMLNode): ParsedXMLNode<string> => {
    const nodeKind = xmlNode.kind;

    const printedAttributes: Partial<ParsedXMlNodePartAttributes<string>> = this.printAttributes(xmlNode.attributes);

    let printedNode: ParsedXMLElement<string> | ParsedXMLComponent<string>;
    if (xmlNode instanceof XMLComponent) {
      printedNode = {
        [nodeKind]: xmlNode.children.map(this.printNode),
        ...printedAttributes,
      } as ParsedXMLComponent<string> & Partial<ParsedXMlNodePartAttributes<string>>;
    } else if (xmlNode instanceof XMLElement) {
      printedNode = {
        [nodeKind]: xmlNode.content,
        ...printedAttributes,
      } as ParsedXMLElement<string> & Partial<ParsedXMlNodePartAttributes<string>>;
    } else {
      throw new Error(`Unknown XML node type: ${xmlNode}`);
    }

    return printedNode;
  };

  printAttributes = (attributes: Record<string, string | number>): Partial<ParsedXMlNodePartAttributes<string>> => {
    const attributesCount = Object.keys(attributes).length;
    if (attributesCount === 0) return {};

    const printedAttributes = Object.entries(attributes).reduce((parsedAttributes, [key, value]) => {
      const attributeName = `@_${key}` as ParsedXMLAttribute<string>;
      return {
        ...parsedAttributes,
        [attributeName]: value.toString(),
      };
    }, {});

    return {
      [PARSED_XML_ATTRIBUTES_KEY]: printedAttributes,
    };
  };
}
