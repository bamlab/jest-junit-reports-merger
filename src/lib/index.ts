import { FileSystem } from './fileSystem';
import { JUnitMerger } from './junitMerger';

export function mergeJunitFiles(output: string, junitFilesGlob: string) {
  const fileSystem = new FileSystem(process.cwd());
  const junitFiles = fileSystem.listFilesPath(junitFilesGlob);

  const junitFileContents = junitFiles.map(fileSystem.readFileContent);

  const junitMerger = new JUnitMerger();
  junitMerger.ingest(junitFileContents);

  fileSystem.writeFileContent(output, junitMerger.print());
}
