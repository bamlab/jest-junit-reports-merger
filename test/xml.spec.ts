import { XMLDocument } from '../src/lib/data/xmlDocument';
import { XMLComponent, XMLElement } from '../src/lib/data/xmlNode';

import { XML } from '../src/lib/xmlModule';

describe('xml', () => {
  it('should parse xml', () => {
    // given
    const junitXml = `<?xml version="1.0" encoding="UTF-8"?>
    <testsuites name="Jest Tests" tests="0" failures="0" errors="0" time="0">
        <testsuite name="jest" tests="0" failures="0" errors="0" time="0" timestamp="2021-03-01T16:00:00">
            <testcase name="test1" classname="test1" time="0.001"/>
        </testsuite>
    </testsuites>`;
    // when
    const xmlModule = new XML();
    const parsedXml = xmlModule.parse(junitXml);
    // then
    expect(parsedXml).toMatchSnapshot();
  });
  it('should print xml', () => {
    // given
    const parsedHeader = new XMLDocument([
      new XMLComponent('?xml', { version: '1.0', encoding: 'UTF-8' }, [new XMLElement('#text', {}, '')]),
    ]);
    // when
    const xmlModule = new XML();
    const printedXML = xmlModule.print(parsedHeader);
    // then
    expect(printedXML).toMatchSnapshot();
  });
});
