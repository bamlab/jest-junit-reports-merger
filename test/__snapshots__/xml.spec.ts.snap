// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xml should parse xml 1`] = `
XMLDocument {
  "nodes": Array [
    XMLComponent {
      "attributes": Object {
        "encoding": "UTF-8",
        "version": "1.0",
      },
      "children": Array [
        XMLElement {
          "attributes": Object {},
          "content": "",
          "kind": "#text",
        },
      ],
      "kind": "?xml",
    },
    XMLComponent {
      "attributes": Object {
        "errors": 0,
        "failures": 0,
        "name": "Jest Tests",
        "tests": 0,
        "time": 0,
      },
      "children": Array [
        XMLComponent {
          "attributes": Object {
            "errors": 0,
            "failures": 0,
            "name": "jest",
            "tests": 0,
            "time": 0,
            "timestamp": "2021-03-01T16:00:00",
          },
          "children": Array [
            XMLComponent {
              "attributes": Object {
                "classname": "test1",
                "name": "test1",
                "time": 0.001,
              },
              "children": Array [],
              "kind": "testcase",
            },
          ],
          "kind": "testsuite",
        },
      ],
      "kind": "testsuites",
    },
  ],
}
`;

exports[`xml should print xml 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>"`;
